### ISSUE TITLE
asciidoctor-diagram generated svg is searched in wrong path 

### ISSUE THREAD

**Issue by frarees** [asciidoctor/asciidoctor-pdf#495]
In my project root I have a Gemfile with: 

```
namespace :book do
    task :build do
        backends = ['html', 'pdf']
        backends.each { |b|
            Asciidoctor.convert_file 'book.adoc',
                to_dir: 'dist/book/' + b,
                mkdirs: true,
                safe: :safe,
                backend: b
        }
    end
end
```

asciidoctor-pdf yields:

```
WARNING: image to embed not found or not readable: <projectdir>/<file>.svg
```

I think it should be resolving the path like this instead `<projectdir>/<to_dir>/<file>.svg`?

The svg is properly generated in `<to_dir>`. 

Thanks.

**Comment by mojavelinux** [comment]
This is one of the challenges we're still trying to reconcile in the Asciidoctor ecosystem. There are a lot of different scenarios to cover and what I've said before is that I think we don't have enough attributes to express it.

When Asciidoctor Diagram runs, it puts images into the imagesoutdir. If not specified, that attribute defaults to the output directory + the imagesdir attribute. That's why the images end up in dist/book/pdf.

However, when Asciidoctor PDF resolves the image, it uses imagesdir + the image target. Since you haven't set imagesdir, that resolves to the same directory as the source file (i.e., the project directory). When an image is resolved, it doesn't know anything about Asciidoctor Diagram and where it is putting images.

One solution to this problem is to set imagesoutdir so that Asciidoctor Diagram generates images into the location where you have all your other images. In other words, set imagesoutdir to match imagesdir.

```
attributes: {
  'imagesdir' => './images'
  'imagesoutdir' => './images'
}
```

Asciidoctor Diagram will still put the cache files in the output directory under .asciidoctor. I think that Asciidoctor Diagram should be enhanced to recognize a cachedir attribute.

Another possible solution to this problem is to introduce a setting in Asciidoctor Diagram that it writes image targets using absolute paths. That way, no matter where the file is generated, Asciidoctor PDF will be able to locate it.

Either way, this is an issue with Asciidoctor Diagram and not Asciidoctor PDF. Asciidoctor PDF can only look for an image where it is told to look for an image. In this case when the output directory is different from the input directory, the path to the generated image node is not resolving to a valid file for embedding (remember, HTML only links to the file, so there are different scenarios at play here).

**Comment by andyjefferson** [comment]
Setting "imagesoutdir" to the source images directory is not an option for most people; the reason being, if you have your docs in such as GitHub checked in, and use Maven to generate the docs under "target/site", having asciidoctor-diagram generate images into the source tree means they could easily be accidentally checked in [although it should be said that this workaround does "work" in terms of generating the PDF and HTML].

Is it an option to treat source images and generated images as distinct? (clearly not needed for the HTML generation). Then the PDF plugin has two "imagesdir"s available to it. It checks in one for the resource, and if not found then checks in the other.

I put a sample project at https://github.com/andyjefferson/asciidoctor-pdf-diagram-combo  but can't find any clean way of getting it to work for the PDF, or maybe I'm missing some trick, in which case feel free to point out my lack of understanding :-)

**Comment by sbrannen** [comment]
This appears to be a duplicate of #271.

**Comment by sbrannen** [comment]
@andyjefferson, there's at least a _work-around_ discussed here: https://github.com/asciidoctor/asciidoctor-pdf/issues/271#issuecomment-166569803. Maybe that works for you, too. Though I haven't personally tried it yet.

**Comment by mojavelinux** [comment]
> Is it an option to treat source images and generated images as distinct? (clearly not needed for the HTML generation). Then the PDF plugin has two "imagesdir"s available to it. It checks in one for the resource, and if not found then checks in the other.

This is might be a direction we need to head in the future. The issue I have with jumping to that now is that it feels like we're adding a solution without totally understanding the problem. Until this can be fully documented, we're really just grasping at straws for a solution. Let's clearly define what has to be done now, then we can imagine a roadmap towards something more flexible.

**Comment by mojavelinux** [comment]
@andyjefferson your sample project definitely helps us along that path. It might be something we want to migrate to the examples project.

**Comment by sbrannen** [comment]
FYI: a feasible solution can now be seen here: https://github.com/junit-team/junit5/commit/cb8d49578740159871a4358707c01050d194e220

**Comment by mojavelinux** [comment]
> Setting "imagesoutdir" to the source images directory is not an option for most people; the reason being, if you have your docs in such as GitHub checked in, and use Maven to generate the docs under "target/site", having asciidoctor-diagram generate images into the source tree means they could easily be accidentally checked in

Actually, Asciidoctor Diagram now generates the images into the build directory by default, so there is no need to set the imagesoutdir attribute or risk checking in generated files. The only requirement is to get all your static images copied to the same location so that all images end up in one place.

> Is it an option to treat source images and generated images as distinct?

One possibility is to first look for an image under imagesdir (when generating a PDF). If the image can't be found, and the imagesoutdir attribute is set and points to a different location, then look there next. That seems like reasonable logic. That way, you would not have to worry about collecting all the images into a single place.

**Comment by mojavelinux** [comment]
> I put a sample project at https://github.com/andyjefferson/asciidoctor-pdf-diagram-combo but can't find any clean way of getting it to work for the PDF

All you are missing is setting imagesdir in the configuration of the PDF converter:

```
<imagesdir>${build.dir}</imagesdir>
```

The Maven plugin already copies all static resources to the output directory by default, so your static images are there. Asciidoctor Diagram now generates into that folder as well. So all images will be located there by the time the converter runs.

You should think of "imagesdir" as "images catalog". This is the base location from where image targets are resolved. The AsciiDoc document doesn't worry about exactly where this file is. It just describes it relative to imagesdir root.

There are tricks you can use to get images to resolve in the preview window, such as on GitHub or in Atom. That's when you define the imagesdir attribute at the top of the document. The value will be overridden when the document is converted by the build tool, so you can safely set it to a value relative to the current document.

**Comment by mojavelinux** [comment]
@sbrannen you almost got it right. You mostly just did some extra stuff you don't need. See https://github.com/junit-team/junit5/pull/779

> We create a custom Copy task that copies all static images to the catalog directory. The asciidoctor task is configured to depend on the new copy task.

Both the Maven and Gradle plugins already handle copying static resources to the output directory, so there's no need for a dedicated task. Asciidoctor Diagram now generates to the output directory by default (outputDir + imagesdir or just imagesdir if the value is absolute).

> Prior to this commit, diagrams created using asciidoctor-diagram were included in the HTML version of the User Guide but not in the PDF version. This is due to a limitation within the Asciidoctor ecosystem. Namely, the PDF backend for Asciidoctor does not know about the fact that asciidoctor-diagram has generated images in a target directory.

I wouldn't call this a limitation so much as a logical challenge. The images need to be where the converter is going to look for them. The converter can only look in one place at the moment (which is really the only limitation). The responsibility of the build tool is to get them there. Asciidoctor Diagram already generates images into the output directory, so the build tool just needs to copy the non-generated ones. And it just so happens it does this by default.

Thus, the only necessary configuration is to give imagesdir an absolute path value when converting to PDF.

**Comment by mojavelinux** [comment]
I am going to consider having Asciidoctor PDF look for images in the imagesoutdir if set and the image cannot otherwise be resolved. I think that would cut this down to 0 configuration (actually, you'd have to define imagesoutdir since we can't know that Asciidoctor Diagram is using a default value).

**Comment by mojavelinux** [comment]
Once my PR is merged into junit5, I think it provides the model that resolves this issue. If we want to have that example brought into the example project for the Maven and Gradle plugins (with PDF output), then I think we should reopen the issue there.

Any other questions?

**Comment by sbrannen** [comment]
Nope. No further questions at this time. Thanks, @mojavelinux!

**Comment by mojavelinux** [comment]
I'm going to mark this as a documentation issue because I think it's worth covering the in README.

**Comment by mojavelinux** [comment]
I have added a note to the README.

Starting in 1.5.0.alpha.17, we're going to force set the data-uri attribute so that Asciidoctor Diagram returns an absolute image path, which Asciidoctor PDF can then locate.
